File: src/main/java/me/testaccount666/serversystem/commands/executables/seen/CommandSeen.java
package me.testaccount666.serversystem.commands.executables.seen;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.userdata.UserManager;
import org.bukkit.command.Command;
import org.bukkit.entity.Player;

import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "seen", tabCompleter = TabCompleterSeen.class)
public class CommandSeen extends AbstractServerSystemCommand {

    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        if (!checkBasePermission(commandSender, "Seen.Use")) return;

        if (arguments.length == 0) {
            general("InvalidArguments", commandSender).syntaxPath(getSyntaxPath(command)).label(label).build();
            return;
        }

        var cachedUserOptional = ServerSystem.Instance.getRegistry().getService(UserManager.class).getUser(arguments[0]);

        if (cachedUserOptional.isEmpty()) {
            ServerSystem.getLog().warning("(CommandSeen) User '${arguments[0]}' is not cached! This should not happen!");
            general("ErrorOccurred", commandSender).label(label).build();
            return;
        }

        var targetUser = cachedUserOptional.get().getOfflineUser();

        if (targetUser.getName().isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[0]).build();
            return;
        }

        var lastSeen = targetUser.getLastSeen();

        if (targetUser instanceof User) lastSeen = System.currentTimeMillis();

        var formattedDate = parseDate(lastSeen);

        command("Seen.SuccessOther", commandSender).target(targetUser.getName().get())
                .postModifier(message -> message.replace("<DATE>", formattedDate)
                        .replace("<IP>", targetUser.getLastKnownIp())).build();
    }


    private String parseDate(long dateMillis) {
        return Instant.ofEpochMilli(dateMillis)
                .atZone(ZoneId.systemDefault())
                .format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm:ss"));
    }

    @Override
    public String getSyntaxPath(Command command) {
        return "Seen";
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        return PermissionManager.hasCommandPermission(player, "Seen.Use", false);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/seen/CommandSeen.kt
package me.testaccount666.serversystem.commands.executables.seen

import me.testaccount666.serversystem.ServerSystem.Companion.log
import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.extensions.*
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.userdata.UserManager
import org.bukkit.command.Command
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

@ServerSystemCommand("seen", [], TabCompleterSeen::class)
class CommandSeen : AbstractServerSystemCommand() {
    override fun minRequiredArguments(command: Command) = 1
    override fun getUsagePermission(command: Command) = "Seen.Use"
    override fun getSyntaxPath(command: Command?) = "Seen"


    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        val cachedUser = getService<UserManager>().getUserOrNull(arguments[0]) ?: run {
            log.warning("(CommandSeen) User '${arguments[0]}' is not cached! This should not happen!")
            commandSender.generalMsg("ErrorOccurred") { label(label) }
            return
        }

        val targetUser = cachedUser.offlineUser.also {
            if (it.getNameOrNull() == null) {
                commandSender.generalMsg("PlayerNotFound") { target(arguments[0]) }
                return
            }
        }

        val lastSeen = if (targetUser is User) System.currentTimeMillis() else targetUser.lastSeen
        val formattedDate = parseDate(lastSeen)

        commandSender.commandMsg("Seen.Success") {
            target(targetUser.getNameOrNull())
            postModifier {
                it.replace("<DATE>", formattedDate)
                    .replace("<IP>", targetUser.lastKnownIp ?: "???")
            }
        }
    }

    private fun parseDate(dateMillis: Long): String {
        return Instant.ofEpochMilli(dateMillis)
            .atZone(ZoneId.systemDefault())
            .format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm:ss"))
    }
}
